New standoff design

benjamingeer opened the pull request:
This pull request is for the implementation of the new standoff design taking shape in #101.

The new standoff design is implemented here mainly in the ontology, which now supports many features that are not yet in the API. The API continues to work as before.

* `StandoffUtil` converts from XML to standoff and back, using its own data structures, without doing anything with RDF.
  * Includes support for non-hierarchical [CLIX](http://conferences.idealliance.org/extreme/html/2004/DeRose01/EML2004DeRose01.html#t6) markup.
* Refactors the knora-base ontology as described [here](https://github.com/dhlab-basel/Knora/issues/101#issuecomment-219990173).
* Maps API v1's standoff tag names onto standoff RDF classes so the API still works.
* Includes UUIDs in standoff tags.
* Generates Knora IRIs for standoff tags, rather than creating them as blank nodes.
* Tests.
* Documentation.

Closes #198.
Closes #199.

tobiasschweizer commented:
TODO:
- abstract out base types for value types (like a date) so they can be inherited be the type specific value objects (like DateValue) and a specific standoff tag (StandoffDateTag).

tobiasschweizer commented on webapi/src/main/scala/org/knora/webapi/responders/v1/ValuesResponderV1.scala:
```diff
@@ -438,6 +442,7 @@ class ValuesResponderV1 extends ResponderV1 {
                                             maybeValuePermissions = permissionsFromDefaults
                                         ).toString()
 
+                                        // append the SPARQL that was generated to create standoff links to the other insert statements
```
@benjamingeer Can this comment be removed? This was my fix to the problem that `standoffLinkInsertSparql` was never used, but now you added it to `allInsertSparql` (line 466). Should the comment be moved there?

benjamingeer replied:
Fixed.

benjamingeer commented:
TODO before we can merge this:

* Generate Knora IRIs for standoff tags, rather than creating them as blank nodes.
* Include UUIDs in standoff tags.

tobiasschweizer commented:
This is the current state of standoff in the `knora-base?  ontology. Most of it is not used yet. At the moment, the application only uses `standoffTagHasStart`, `standoffTagHasEnd` (if appropriate: `knora-base:standoffTagHasLink` or `knora-base:standoffTagHasHref`) when creating a standoff tag and assigns its IRI as the `rdf:type`.

```turtle
#################################################################
#
# Generic Standoff Tag
#
#################################################################

###  http://www.knora.org/ontology/knora-base#StandoffTag

:StandoffTag rdf:type owl:Class ;

          rdfs:subClassOf [ rdf:type owl:Restriction ;
                            owl:onProperty :standoffTagHasStart ;
                            owl:cardinality "1"^^xsd:nonNegativeInteger
                          ] ,
                          [ rdf:type owl:Restriction ;
                            owl:onProperty :standoffTagHasEnd ;
                            owl:cardinality "1"^^xsd:nonNegativeInteger
                          ] ,
                          [ rdf:type owl:Restriction ;
                            owl:onProperty :standoffTagHasUUID ;
                            owl:maxCardinality "1"^^xsd:nonNegativeInteger
                          ],
                          [ rdf:type owl:Restriction ;
                            owl:onProperty :standoffTagHasIndex ;
                            owl:maxCardinality "1"^^xsd:nonNegativeInteger
                          ],
                          [ rdf:type owl:Restriction ;
                            owl:onProperty :standoffTagHasParentIndex ;
                            owl:maxCardinality "1"^^xsd:nonNegativeInteger
                          ],
                          [ rdf:type owl:Restriction ;
                            owl:onProperty :standoffTagHasTagname ;
                            owl:maxCardinality "1"^^xsd:nonNegativeInteger
                          ],
                          [ rdf:type owl:Restriction ;
                            owl:onProperty :standoffTagHasXMLNamespace ;
                            owl:maxCardinality "1"^^xsd:nonNegativeInteger
                          ] ,
                          [ rdf:type owl:Restriction ;
                            owl:onProperty :standoffTagHasXMLAttribute ;
                            owl:minCardinality "0"^^xsd:nonNegativeInteger
                          ] ;

          rdfs:comment "Represents a standoff tag in a TextValue"@en .


### http://www.knora.org/ontology/knora-base#StandoffXMLAttribute

:StandoffXMLAttribute rdf:type owl:Class ;

           rdfs:subClassOf [ rdf:type owl:Restriction ;
                             owl:onProperty :standoffXMLAttributeHasXMLNamespace ;
                             owl:maxCardinality "1"^^xsd:nonNegativeInteger
                           ],
                           [ rdf:type owl:Restriction ;
                             owl:onProperty :standoffXMLAttributeHasName ;
                             owl:cardinality "1"^^xsd:nonNegativeInteger
                           ],
                           [ rdf:type owl:Restriction ;
                             owl:onProperty :standoffXMLAttributeHasValue ;
                             owl:cardinality "1"^^xsd:nonNegativeInteger
                           ] ;

          rdfs:comment "Represents an attribute of a standoff tag"@en .


#################################################################
#
# Intermediate Classes for different Types of Standoff Tags
#
#################################################################

###  http://www.knora.org/ontology/knora-base#StandoffVisualTag

:StandoffVisualTag rdf:type owl:Class ;

               rdfs:subClassOf :StandoffTag ;

               rdfs:comment "Represents visual markup information in a TextValue"@en .


###  http://www.knora.org/ontology/knora-base#StandoffStructuralTag

:StandoffStructuralTag rdf:type owl:Class ;

               rdfs:subClassOf :StandoffTag ;

               rdfs:comment "Represents structural markup information in a TextValue"@en .


###  http://www.knora.org/ontology/knora-base#StandoffReferenceTag

:StandoffReferenceTag rdf:type owl:Class ;

               rdfs:subClassOf :StandoffTag ;

               rdfs:comment "Represents references in a TextValue"@en .


###  http://www.knora.org/ontology/knora-base#StandoffDataTypeTag

:StandoffDataTypeTag rdf:type owl:Class ;

                 rdfs:subClassOf :StandoffTag ;

                 rdfs:comment "Represents a knora-base value type in a TextValue"@en .


#################################################################
#
# Standoff Tags that can be instantiated
#
#################################################################


###  http://www.knora.org/ontology/knora-base#StandoffHrefTag

:StandoffHrefTag rdf:type owl:Class ;

              rdfs:subClassOf :StandoffReferenceTag ,
                              [ rdf:type owl:Restriction ;
                                owl:onProperty :standoffTagHasHref ;
                                owl:cardinality "1"^^xsd:nonNegativeInteger
                              ] ;

              rdfs:comment "Represents a hyperlink in a TextValue"@en .



###  http://www.knora.org/ontology/knora-base#StandoffLinkTag

:StandoffLinkTag rdf:type owl:Class ;

              rdfs:subClassOf :StandoffReferenceTag ,
                              [ rdf:type owl:Restriction ;
                                owl:onProperty :standoffTagHasLink ;
                                owl:cardinality "1"^^xsd:nonNegativeInteger
                              ] ;

              rdfs:comment "Represents a reference to a Knora resource in a TextValue"@en .




###  http://www.knora.org/ontology/knora-base#StandoffDateValueTag

:StandoffDateValueTag rdf:type owl:Class ;

                        rdfs:subClassOf :StandoffDataTypeTag,
                                        :DateBase ;

                        rdfs:comment "Represents a DateValue in a TextValue"@en .


###  http://www.knora.org/ontology/knora-base#StandoffParagraphTag

:StandoffParagraphTag rdf:type owl:Class ;

                rdfs:subClassOf :StandoffStructuralTag ;

                rdfs:comment "Represents a paragraph in a TextValue"@en .

###  http://www.knora.org/ontology/knora-base#StandoffHeader1Tag

:StandoffHeader1Tag rdf:type owl:Class ;

                rdfs:subClassOf :StandoffStructuralTag ;

                rdfs:comment "Represents a header of level 1 in a TextValue"@en .


###  http://www.knora.org/ontology/knora-base#StandoffHeader2Tag

:StandoffHeader2Tag rdf:type owl:Class ;

                rdfs:subClassOf :StandoffStructuralTag ;

                rdfs:comment "Represents a header of level 2 in a TextValue"@en .


###  http://www.knora.org/ontology/knora-base#StandoffHeader3Tag

:StandoffHeader3Tag rdf:type owl:Class ;

                rdfs:subClassOf :StandoffStructuralTag ;

                rdfs:comment "Represents a header of level 3 in a TextValue"@en .


###  http://www.knora.org/ontology/knora-base#StandoffHeader4Tag

:StandoffHeader4Tag rdf:type owl:Class ;

                rdfs:subClassOf :StandoffStructuralTag ;

                rdfs:comment "Represents a header of level 4 in a TextValue"@en .


###  http://www.knora.org/ontology/knora-base#StandoffHeader5Tag

:StandoffHeader5Tag rdf:type owl:Class ;

                rdfs:subClassOf :StandoffStructuralTag ;

                rdfs:comment "Represents a header of level 3 in a TextValue"@en .


###  http://www.knora.org/ontology/knora-base#StandoffHeader6Tag

:StandoffHeader6Tag rdf:type owl:Class ;

                rdfs:subClassOf :StandoffStructuralTag ;

                rdfs:comment "Represents a header of level 3 in a TextValue"@en .


###  http://www.knora.org/ontology/knora-base#StandoffOrderedListTag

:StandoffOrderedListTag rdf:type owl:Class ;

                rdfs:subClassOf :StandoffStructuralTag ;

                rdfs:comment "Represents an ordered list in a TextValue"@en .


###  http://www.knora.org/ontology/knora-base#StandoffUnorderedListTag

:StandoffUnorderedListTag rdf:type owl:Class ;

                rdfs:subClassOf :StandoffStructuralTag ;

                rdfs:comment "Represents an unordered list in a TextValue"@en .


###  http://www.knora.org/ontology/knora-base#StandoffListElementTag

:StandoffListElementTag rdf:type owl:Class ;

                rdfs:subClassOf :StandoffStructuralTag ;

                rdfs:comment "Represents a list element in a TextValue"@en .


###  http://www.knora.org/ontology/knora-base#StandoffStyleTag

:StandoffStyleTag rdf:type owl:Class ;

                rdfs:subClassOf :StandoffVisualTag ;

                rdfs:comment "Represents a style in a TextValue"@en .


###  http://www.knora.org/ontology/knora-base#StandoffItalicTag

:StandoffItalicTag rdf:type owl:Class ;

                rdfs:subClassOf :StandoffVisualTag ;

                rdfs:comment "Represents italics in a TextValue"@en .


###  http://www.knora.org/ontology/knora-base#StandoffBoldTag

:StandoffBoldTag rdf:type owl:Class ;

                rdfs:subClassOf :StandoffVisualTag ;

                rdfs:comment "Represents bold text in a TextValue"@en .


###  http://www.knora.org/ontology/knora-base#StandoffUnderlineTag

:StandoffUnderlineTag rdf:type owl:Class ;

                rdfs:subClassOf :StandoffVisualTag ;

                rdfs:comment "Represents underlined text in a TextValue"@en .


###  http://www.knora.org/ontology/knora-base#StandoffStrikethroughTag

:StandoffStrikethroughTag rdf:type owl:Class ;

                rdfs:subClassOf :StandoffVisualTag ;

                rdfs:comment "Represents struck text in a TextValue"@en .


###  http://www.knora.org/ontology/knora-base#StandoffSuperscriptTag

:StandoffSuperscriptTag rdf:type owl:Class ;

                rdfs:subClassOf :StandoffVisualTag ;

                rdfs:comment "Represents superscript in a TextValue"@en .


###  http://www.knora.org/ontology/knora-base#StandoffSubscriptTag

:StandoffSubscriptTag rdf:type owl:Class ;

                rdfs:subClassOf :StandoffVisualTag ;

                rdfs:comment "Represents subscript in a TextValue"@en .

```

benjamingeer commented:
Removed XML namespaces and XML attributes from `knora-base`, because these will be handled by the XML-to-standoff mapping, which should work like this.

### XML

```
<n1:person n2:archiveID="123" n2:unsure="true">Bernoulli</n1:person>
```

### Mapping

```
n1:person -> myproject:StandoffPersonTag

n2:archiveID -> myproject:hasArchiveID (type string)

n2:unsure -> myproject:unsure (type boolean)
```

### RDF

```
<http://data.knora.org/a/values/b/1> a myproject:StandoffPersonTag ;
    myproject:hasArchiveID "123" ;
    myproject:unsure "true"^^xsd:boolean .
```

benjamingeer commented:
In the standoff transformation, what do I do about this?

```
_:node1ap8nhe1mx1 knora-base:standoffHasAttribute "linebreak" ;
	knora-base:standoffHasStart 4 ;
	knora-base:standoffHasEnd 4 ;
	a knora-base:StandoffVisualAttribute .
```

There is no `linebreak` in the `StandoffTagV1` enumeration.

tobiasschweizer commented:
You can delete linebreaks in standoff (like we did manually in the test data). But we have to represent them in the string instead.

benjamingeer commented:
My suggestion was here: https://github.com/dhlab-basel/Knora/issues/101#issuecomment-234352467

tobiasschweizer commented:
yes, I know that we talked about it :-)

benjamingeer commented:
I think the real problem for us is that `\r\n`, used on Windows machines, is two characters long. What I'm worrying about is if we get text with standoff from a Windows machine, containing `\r\n`. Such text could be in a TTL file or in JSON that we receive over the API. If we convert `\r\n` to a single character (whether it's `\n` or something else), the standoff positions will be wrong.

How can we ensure that newlines will always be represented by a single character in text?

Also, in the data transformer, if I handle a `linefeed` standoff tag by adding a linefeed to the text, the standoff positions will also be wrong.

benjamingeer commented:
After discussion:

* The JavaScript in SALSAH currently uses `\n` for newline; this is fine. It also adds `\r` after block tags to avoid text running together. This is OK, but we should use another non-printing delimiter character instead, like [INFORMATION SEPARATOR TWO](http://www.fileformat.info/info/unicode/char/001e/index.htm). Therefore the JavaScript needs to be changed to do this, and the data transformer needs to replace `\r` with `INFORMATION SEPARATOR TWO`.
* The data transformer has to remove each `linebreak` standoff tag, insert `\n` in its place, and increment the subsequent standoff positions by one.
* Input validation should throw `BadRequestException` if we get text containing `\r` in an API request.
* If someone makes a Turtle file containing `\r\n` on Windows and imports it directly into the triplestore, there isn't much we can do except write documentation warning people not to do that. We could make an API route for uploading a Turtle file, which would enable us to check for that. We could also check for `\r` when reading a text value, and throw `InconsistentTriplestoreDataException`.

tobiasschweizer commented:
In the GUI, the JS `jquery.htmleditor.js` converting from and to standoff has to be adapted (as mentioned above in https://github.com/dhlab-basel/Knora/pull/166#issuecomment-242702827):

in `__convert_html2lin`:

```javascript
if (cur_node.nodeName.toLowerCase() == 'p' || cur_node.nodeName.toLowerCase() == 'li' || cur_node.nodeName.toLowerCase()[0] == 'h') {
						// it is a block element, insert a \r later so that in utf8str their string representations are not attached to each other
						control_chars.push({pos: char_count, 'char': '\r'});
						char_count++; // increment by one for '\r' to be inserted
						
						// the children have been processed (we know the length now), add the prop
						add_prop(localdata.settings.matching[cur_node.nodeName.toLowerCase()], tmp_char_count, char_count - tmp_char_count);
					}
```

and `convert_lin2html`:

```javascript

// replace '\n' with <br>
		html = html.replace(/\n/g, '<br/>');
		// remove \r since they are represented by block elements
		html = html.replace(/\r/g, '');


```

tobiasschweizer commented:
my compliments for the transformation script: I ran it with the beol data three times: permissions, strings (valueHasString) and standoff.

It works! :-)

benjamingeer commented:
Great! I should make it more robust, I think it can crash if your data doesn't need the transformation you request.

tobiasschweizer commented:
Let's do that next week! 

I am getting traumatized by XML and its handling of newlines and spaces ...

benjamingeer commented:
I added the features described above to `TransformData`. I also found that there was still a lot of test data containing a misspelled old-style permission, `hasChangeRightsPermisson` (#62), mixed together with new-style `hasPermissions` properties generated by `TransformData` (which hadn't recognised the misspelled property). To fix this, I modified `TransformData` to recognise the misspelled property as equivalent to the correctly-spelled one, and ran it on the most recent versions of the test data files that still had the old-style permissions. The result seems OK; the tests pass.

`TransformData` now also has a `-t all` option to run all transformations. Running it on correct data now has no effect.

benjamingeer commented:
I think `StandoffHrefTag` should be a subclass of `UriBase`, and should be renamed `StandoffUriTag`.

Also, why do we have `ColorBase`, `IntBase`, `DecimalBase`, `BooleanBase`, and `IntervalBase`, but no `StandoffColorValueTag`, etc.? If we aren't going to use these `*Base` classes to make standoff tag classes, then I think we don't need them in the ontology.
